Mathlib/Analysis/Complex/HasPrimitives.lean

```diff
@@ -0,0 +1,492 @@
+satisfy this property.) To prove Morera's theorem, we first define the `WedgeInt`, which is the
+integral of a function over a "wedge" (a horizontal segment followed by a vertical segment in the
+disc), and compute its derivative.
+
+## Main results
+
+* `VanishesOnRectanglesInDisc.diff_of_wedges`: If a function `f` vanishes on all rectangles in a
+  disc with center `c`, then the wedge integral from `c` to `w` minus the wedge integral from
+  `c` to `z` is equal to the wedge integral from `z` to `w`.
+
+* `deriv_of_wedgeInt`: The derivative of the wedge integral is the function being integrated.
+
+* `VanishesOnRectanglesInDisc.hasPrimitive`: **Morera's Theorem**: A function which is
+  continuous on a disc and whose integral on rectangles in the disc vanishes has a primitive
+  on the disc (defined by the wedge integral).
+
+* `hasPrimitives_on_disc`: A holomorphic function on a disc has primitives.
+
+## Tags
+  Holomorphic functions, primitives
+
+TODO: Extend to holomorphic functions on simply connected domains. (In particular, this allows one
+to define the complex logarithm of a nonvanishing function on a simply connected domain.)
+-/
+
+open Complex Topology Set Metric
+
+set_option autoImplicit true
+
+open scoped Interval
+
+namespace Complex
+
+section Asymptotics
+
+/-- As `w → z`, `w.re - z.re` is big-O of `w - z`. -/
+lemma re_isBigO {z : ℂ} : (fun (w : ℂ) ↦ w.re - z.re) =O[𝓝 z] fun w ↦ w - z := by
+  rw [Asymptotics.isBigO_iff]
+  use 1
+  filter_upwards
+  intro w
+  simp only [Real.norm_eq_abs, Complex.norm_eq_abs, one_mul]
+  rw [← Complex.sub_re]
+  exact Complex.abs_re_le_abs (w - z)
+
+/-- As `w → z`, `w.im - z.im` is big-O of `w - z`. -/
+lemma im_isBigO {z : ℂ} : (fun (w : ℂ) ↦ w.im - z.im) =O[𝓝 z] fun w ↦ w - z := by
+  rw [Asymptotics.isBigO_iff]
+  use 1
+  filter_upwards
+  intro w
+  simp only [Real.norm_eq_abs, Complex.norm_eq_abs, one_mul]
+  rw [← Complex.sub_im]
+  exact Complex.abs_im_le_abs (w - z)
+
+end Asymptotics
+
+section reProdIm
+
+/-- The preimage under `equivRealProd` of `s ×ˢ t` is `s ×ℂ t`. -/
+lemma preimage_equivRealProd_prod (s t : Set ℝ) : equivRealProd ⁻¹' (s ×ˢ t) = s ×ℂ t := rfl
+
+/-- The inequality `s × t ⊆ s₁ × t₁` holds in `ℂ` iff it holds in `ℝ × ℝ`. -/
+lemma reProdIm_subset_iff {s s₁ t t₁ : Set ℝ} : s ×ℂ t ⊆ s₁ ×ℂ t₁ ↔ s ×ˢ t ⊆ s₁ ×ˢ t₁ := by
+  rw [← @preimage_equivRealProd_prod s t, ← @preimage_equivRealProd_prod s₁ t₁]
+  exact Equiv.preimage_subset equivRealProd _ _
+
+/-- If `s ⊆ s₁ ⊆ ℝ` and `t ⊆ t₁ ⊆ ℝ`, then `s × t ⊆ s₁ × t₁` in `ℂ`. -/
+lemma reProdIm_subset_iff' {s s₁ t t₁ : Set ℝ} :
+    s ×ℂ t ⊆ s₁ ×ℂ t₁ ↔ s ⊆ s₁ ∧ t ⊆ t₁ ∨ s = ∅ ∨ t = ∅ := by
+  convert prod_subset_prod_iff
+  exact reProdIm_subset_iff
+
+end reProdIm
+
+section Rectangle
+
+/-- A `Rectangle` is an axis-parallel rectangle with corners `z` and `w`. -/
+def Rectangle (z w : ℂ) : Set ℂ := [[z.re, w.re]] ×ℂ [[z.im, w.im]]
```

Comment:
If we introduce this definition, then it should be near `reProdIm` (if `uIcc` is available there), and existing code should be migrated to it.

Reply:
ok, attempting refactor...